Enhance the existing create account example

SimiHunjan opened the pull request:
Draft PR for account create example updates.

SimiHunjan committed:
Update and rename script-account.js to createAccount.js

Enhance the existing example.

- Remove the use of extra account variables
- Added key generation for the new account
- Replaced the private key in setKey() for account creation with the public key
- Removed "bguiz" and updated with "enterYourName" from YOUR_NAME
- Added console log statements to separate a grouping of steps in the console to make it easier on the eyes to read and copy
- Included to the steps where we get mirror node data and HashScan URLs right after the specific transactions are executed. When you are following the tutorial you will want to talk about these right after the transaction is executed

SimiHunjan committed:
Update createAccount.js

SimiHunjan committed:
Update createAccount.js

SimiHunjan committed:
Update createAccount.js

add mirror node create account transaction request URL

SimiHunjan committed:
Update createAccount.js

update the value for 10 to 6 for the timeout

SimiHunjan commented:
The output:
![Screenshot 2024-07-10 at 9 22 25 PM](https://github.com/hedera-dev/hello-future-world-js/assets/21989138/e40ca36a-290c-4d33-9826-b788d5fea8bb)

SimiHunjan committed:
Update createAccount.js

add comments to explain the reformatting of the transaction ID to be accepted  in the mirror node query params

SimiHunjan committed:
Update createAccount.js

- Added getting the status of the receipt for the two transactions
- Added balance check after the transfer to show the updated balance of the new account

SimiHunjan committed:
Update createAccount.js

Updated console log statement

SimiHunjan committed:
Update createAccount.js

Update console log statement to indicate user should copy and paste the URL for HashScan to their browser

SimiHunjan commented:
With the late updates the output should look like the following:

```
🔵 Creating, signing and submitting the account create transaction...
The account create transaction ID: 0.0.1455@1720736194.924048981
The account create status: SUCCESS
The new account ID:  0.0.4524624
The new account balance is: 5 ℏ
 
🔵 Get account data from the Hedera Mirror Node...
The account info Hedera Mirror Node API request: 
https://testnet.mirrornode.hedera.com/api/v1/accounts/0.0.4524624?limit=1&order=asc&transactiontype=cryptotransfer&transactions=false
The account memo:  Hello Future World from YourNameHere's first account!
 
The account create transaction Hedera Mirror Node API request: 
 https://testnet.mirrornode.hedera.com/api/v1/transactions/0.0.1455-1720736194-924048981?nonce=0
 
🔵 View the account on HashScan...
Copy and paste this URL in your browser:  https://hashscan.io/testnet/account/0.0.4524624
 
🔵 View the account create transaction on HashScan...
Copy and paste this URL in your browser:  https://hashscan.io/testnet/transaction/0.0.1455@1720736194.924048981
 
🔵 Creating, signing, and submitting the transfer transaction...
The transfer transaction ID: 0.0.1455@1720736204.225786879
The transfer transaction status is: SUCCESS
The new account balance after the transfer; 10 ℏ
 
🔵 Get transfer transaction data from the Hedera Mirror Node...
The transfer transaction Hedera Mirror Node API request: 
 https://testnet.mirrornode.hedera.com/api/v1/transactions/0.0.1455-1720736204-225786879?nonce=0
 
The transfer transaction transfer amounts in hbar 
 [ 0.00014523, -5.0015137, 5 ]
 
🔵 View the transfer transaction transaction in HashScan...
Copy and paste this URL in your browser:  https://hashscan.io/testnet/transaction/0.0.1455@1720736204.225786879
```

SimiHunjan committed:
Update createAccount.js

update console log statement to describe the transfer transaction amounts in more detail

SimiHunjan commented:
After the last commit:

```
🔵 Creating, signing and submitting the account create transaction...
The account create transaction ID: 0.0.1455@1720736420.147092305
The account create status: SUCCESS
The new account ID:  0.0.4524641
The new account balance is: 5 ℏ
 
🔵 Get account data from the Hedera Mirror Node...
The account info Hedera Mirror Node API request: 
https://testnet.mirrornode.hedera.com/api/v1/accounts/0.0.4524641?limit=1&order=asc&transactiontype=cryptotransfer&transactions=false
The account memo:  Hello Future World from YourNameHere's first account!
 
The account create transaction Hedera Mirror Node API request: 
 https://testnet.mirrornode.hedera.com/api/v1/transactions/0.0.1455-1720736420-147092305?nonce=0
 
🔵 View the account on HashScan...
Copy and paste this URL in your browser:  https://hashscan.io/testnet/account/0.0.4524641
 
🔵 View the account create transaction on HashScan...
Copy and paste this URL in your browser:  https://hashscan.io/testnet/transaction/0.0.1455@1720736420.147092305
 
🔵 Creating, signing, and submitting the transfer transaction...
The transfer transaction ID: 0.0.1455@1720736426.093702245
The transfer transaction status is: SUCCESS
The new account balance after the transfer; 10 ℏ
 
🔵 Get transfer transaction data from the Hedera Mirror Node...
The transfer transaction Hedera Mirror Node API request: 
 https://testnet.mirrornode.hedera.com/api/v1/transactions/0.0.1455-1720736426-093702245?nonce=0
 
The transaction fee , debit and credit amounts of the transfer transaction 
 [ 0.00014523, -5.0015137, 5 ]
 
🔵 View the transfer transaction transaction in HashScan...
Copy and paste this URL in your browser:  https://hashscan.io/testnet/transaction/0.0.1455@1720736426.093702245
 
```

bguiz committed:
proj: rename script file to enable sensible git diff

Signed-off-by: Brendan Graetz <bguiz@users.noreply.github.com>

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
-    const YOUR_NAME = 'bguiz';
+    const YOUR_NAME = '<enterYourName>';
+
+    console.log("\x1b[34m%s\x1b[0m","🔵 Creating, signing and submitting the account create transaction...");
```
I like adding text colour + emojis to make the output easier to parse!

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
-    // NOTE: Create new account using AccountCreateTransaction
+    //Generate a key for your new account
+    const account1PrivateKey = PrivateKey.generateECDSA();
+    const account1PublicKey = account1PrivateKey.publicKey;
```
I think we can remove this since it's not being used.

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+    console.log("The new account balance is: " + hbarBalance);
+    console.log(" ");
+
+    console.log("\x1b[34m%s\x1b[0m","🔵 Get account data from the Hedera Mirror Node...");
```
For the terminal ANSI colours, we'll need to detect and disable in terminals which don't have this feature. Let me look into this.

Also, let's use a var/ const for these instances.

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+        "\x1b[34m%s\x1b[0m",
+        "🔵 View the account create transaction on HashScan..."
+      );
+    const accountCreateTxVerifyHashscanUrl = `https://hashscan.io/testnet/transaction/${accountCreateTransactionId}`;
```
Probably do not need the Hashscan URLs for both the account and the account create transactions. Since this is getting started/ hello world, we're going for the absolute minimum to demo this area, rather than for a comprehensive tutorial.

Since there'll be a Hashscan URL for another transaction later (for transfer transaction), that box is checked anyway - so I vote to drop this.

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+
+    const transferTx = await new TransferTransaction()
+        //Debit 5 hbars from the operator account
+        .addHbarTransfer(operatorId, -5)
```
Note that I was avoiding the use of `TransferTransaction.addHbarTransfer` because it appears to be missing from the reference documentation. See: https://github.com/hashgraph/hedera-docs/issues/727

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+        //Debit 5 hbars from the operator account
+        .addHbarTransfer(operatorId, -5)
+         //Credit 5 hbars from the new account
+        .addHbarTransfer(account1Id, 5)
```
Also note that the amounts original chosen were a reference to Planck's constant, whose unicode symbol is the same as that of Hbar (hence the non-round number).

Was intending this as an easter egg of sorts 😉

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
-    })
+    console.log("\x1b[34m%s\x1b[0m", "🔵 Creating, signing, and submitting the transfer transaction...");
+
+    const transferTx = await new TransferTransaction()
```
Note that previously, we had 1 debit + 2 credits in the `TransferTransaction`, and now it is 1 debit + 1 credit.

Something to consider, as this is something we can flex: The support for multiple transfers in a single transaction, atomically, and natively built into the protocol. (This a hard-to-solve problem, with only 3rd-party/ BYO solutions, in other EVM networks.)

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+    );
+    console.log(" ");
+
+    //The transfer transaction assessed transaction fee, debits, and credits in HBAR
```
Thanks for adding all of these comments BTW, I think it makes it much easier for a developer to follow along and understand the SDK methods.

I'll try to stick to this level of comments in the subsequent HW seqs.

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
-scriptAccount();
+    const transferJsonAccountTransfers = transferJson?.transactions[0]?.transfers;
+    const transferJsonAccountTransfersFinal3Amounts = transferJsonAccountTransfers
+        ?.slice(-3)
```
If we're doing 1 debit + 1 credit in the `TransferTransaction`, then this should be just the final 2 amounts (instead of 3).

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+    const transferJsonAccountTransfers = transferJson?.transactions[0]?.transfers;
+    const transferJsonAccountTransfersFinal3Amounts = transferJsonAccountTransfers
+        ?.slice(-3)
+        ?.map((obj) => obj.amount / 100000000);
```
Perhaps use something in `HbarUnit` instead of the literal `100000000` here? I has been imported already.
... and I see that it it now an unused import.

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+        "\x1b[34m%s\x1b[0m",
+        "🔵 View the account create transaction on HashScan..."
+    );
+    const accountCreateTxVerifyHashscanUrl = `https://hashscan.io/testnet/transaction/${accountCreateTransactionId}`;
+    console.log(
+        "Copy and paste this URL in your browser: ",
+        accountCreateTxVerifyHashscanUrl
+    );
+    console.log(" ");
```
This seems to be repeated

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
-    const accountCreateJsonAccountMemo = accountCreateJson?.memo;
-    console.log('accountCreateJsonAccountMemo:', accountCreateJsonAccountMemo);
+    //Get the new account balance from a consensus node
+    const newAccountBalance = new AccountBalanceQuery()
```
This should be imported

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
-    const operatorId = AccountId.fromString(operatorIdStr);
-    const operatorKey = PrivateKey.fromStringECDSA(operatorKeyStr);
+    const operatorAccountId = AccountId.fromString(operatorIdStr);
+    const operatorEcdsaKey = PrivateKey.fromStringECDSA(operatorKeyStr);
```
This was created, but not used when initialising the `Client` object.
It appears to be a rename to include "ecdsa" without any functional difference.
Assume that it is OK to revert to original, so that it runs.

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
+        transferTxVerifyHashscanUrl
+    );
+    console.log(" ");
+    client.close();
```
`client.close` can be called earlier, after the final SDK method

bguiz commented on account/script-account.js:
```diff
@@ -11,105 +11,212 @@ import {
-    client.close();
+    //Get the transfer transaction receipt
+    const transferTxReceipt = await transferTxSubmitted.getReceipt(client);
+    const transactionStatus = transferTxReceipt.status;
```
Great - thanks for adding this!

bguiz committed:
fix: for misc items in own PR review

- see https://github.com/hedera-dev/hello-future-world-js/pull/3#pullrequestreview-2173840551

Signed-off-by: Brendan Graetz <bguiz@users.noreply.github.com>
